### PR TITLE
types: refine the parsing logic of INTERVAL to correct DATE_ADD/DATE_SUB

### PR DESCRIPTION
**Pull request by b41sh**
### What problem does this PR solve? 

fix [#9715](https://github.com/pingcap/tidb/issues/9715), the problem that `date_add` and `date_sub` are incompatible with MySQL.

### What is changed and how it works?

modify `extractDayHour` and `extractYearMonth` function to compatible with Mysql

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -1753,41 +1753,82 @@ func extractDayMinute(format string) (int64, int64, int64, float64, error) {
...
+// MySQL permits any punctuation delimiter in the expr format. it may start with a - for negative intervals
 func extractDayHour(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	if len(format) == 0 {
+		return 0, 0, 0, 0, nil
 	}
+	neg := false
+	format0 := format
```
Comment (zz-jason) [inline types/time.go]: s/format0/originalFmt/

File: types/time.go [hunk 2]
```diff
@@ -1753,41 +1753,82 @@ func extractDayMinute(format string) (int64, int64, int64, float64, error) {
...
 
-	days, err := strconv.ParseInt(fields[0], 10, 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	if len(fields) == 0 {
+		return 0, 0, 0, 0, nil
+	} else if len(fields) > 2 {
+		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format0)
+	}
+	field0 := "0"
```
Comment (zz-jason) [inline types/time.go]: s/field0/fieldDays/

File: types/time.go [hunk 3]
```diff
@@ -1753,41 +1753,82 @@ func extractDayMinute(format string) (int64, int64, int64, float64, error) {
...
+	if len(fields) > 1 {
+		field0 = fields[len(fields)-2]
 	}
+	field1 := fields[len(fields)-1]
```
Comment (zz-jason) [inline types/time.go]: s/field1/fieldHours/

File: types/time.go [hunk 4]
```diff
@@ -1753,41 +1753,82 @@ func extractDayMinute(format string) (int64, int64, int64, float64, error) {
...
+		hours *= -1
 	}
-
 	return 0, 0, days, hours * float64(gotime.Hour), nil
 }
 
-// extractYearMonth extracts year and month from a string and its format is `YYYY-MM`.
+// extractYearMonth extracts year and month from a string and its suggested format is `YYYY-MM`.
+// MySQL permits any punctuation delimiter in the expr format. it may start with a - for negative intervals
 func extractYearMonth(format string) (int64, int64, int64, float64, error) {
```
Comment (zz-jason) [inline types/time.go]: looks like the main logic of `extractDayHour` and `extractYearMonth` is the same? Can we extract a function and reuse these codes?

Comment (b41sh) [reply types/time.go]: @zz-jason 
I use extract a function to handle all time format.
PTAL

File: types/time.go [hunk 5]
```diff
@@ -1563,234 +1607,67 @@ func extractSingleTimeValue(unit string, format string) (int64, int64, int64, fl
...
-	}
-
-	minutes, err := strconv.ParseFloat(fields[1], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
+// extractTimeValue extracts years, months, days, microseconds from a string
+// MySQL permits any punctuation delimiter in the expr format.
+// See https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals
+func extractTimeValue(format string, index, cnt int) (int64, int64, int64, float64, error) {
```
Comment (qw4990) [inline types/time.go]: This useless blank line can be removed.

Comment (b41sh) [reply types/time.go]: already deleted

File: types/time.go [hunk 6]
```diff
@@ -1563,234 +1607,66 @@ func extractSingleTimeValue(unit string, format string) (int64, int64, int64, fl
...
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+// extractTimeValue extracts years, months, days, microseconds from a string
+// MySQL permits any punctuation delimiter in the expr format.
+// See https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals
+func extractTimeValue(format string, index, cnt int) (int64, int64, int64, float64, error) {
+	neg := false
+	originalFmt := format
+	if len(format) > 0 && format[0] == '-' {
```
Comment (qw4990) [inline types/time.go]: How about trim prefix of this string, `"- 1:1"` and `" - 1:1"` will be both recognized as negative in MySQL.

Comment (b41sh) [reply types/time.go]: add trim space
PTAL @qw4990

**Comment by b41sh** [comment]
@zz-jason PTAL

File: types/time.go [hunk 7]
```diff
@@ -1563,234 +1607,67 @@ func extractSingleTimeValue(unit string, format string) (int64, int64, int64, fl
...
-	return 0, 0, days, value, nil
-}
-
-// extractDaySecond extracts day and hour from a string and its format is `DD HH:MM:SS`.
-func extractDaySecond(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	matches := numericRegex.FindAllString(format, -1)
+	if len(matches) > cnt {
```
Comment (zz-jason) [inline types/time.go]: should it be `len(matches) != cnt` ?

Comment (b41sh) [reply types/time.go]: MySQL allows missing left parts of interval value. For example, if you specify a unit of DAY_SECOND, the value of expr is expected to have days, hours, minutes, and seconds parts. If you specify a value like '1:10', MySQL assumes that the days and hours parts are missing and the value represents minutes and seconds.

All of he following input are valid:

select date_add("2019-01-01 00:00:00",INTERVAL "" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1 1" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1 1 1" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1 1 1 1" DAY_SECOND);

Take a look at this document
[Temporal Intervals](https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals)

Comment (zz-jason) [reply types/time.go]: okay

**Comment by zz-jason** [comment]
@b41sh please cherry pick this PR to release-2.1

**Comment by b41sh** [comment]
ok

**Pull request merged**